core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -105,14 +100,13 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                     switch (currentFieldName) {
                     case "must":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            mustClauses.add(query);
-                        }
+                        mustClauses.add(query);
                         break;
                     case "should":
                         query = parseContext.parseInnerQueryBuilder();
-                        if (query != null) {
-                            shouldClauses.add(query);
+                        shouldClauses.add(query);
+                        // EmptyQueryBuilder does not add lucene query later, skip setting minuminShouldMatch
+                        if (query != EmptyQueryBuilder.PROTOTYPE) {
```

Comment:
other question, sorry! but I still find this confusing... previously we had null here for two situations:
1) empty filter/query
2) queries that after getting parsed would become a null query (their parse method returns null)

the first case falls now under the EmptyQueryBuilder case, which we handle here. But there can still be queries whose toQuery returns null, which need to be ignored, but that we know only in the toQuery, but we still set the minimumShouldMatch only based on case 1). I think setting the minimumShouldMatch like should happen later in the toQuery, which kinda makes sense, it needs to be done only on the resulting lucene query, not against the query builder.

So we can potentially add empty query here to the clauses as long as we move this check for null should clauses only in the toQuery method, if that makes sense